### PR TITLE
[rfc][draft] Stable autoscaler interface

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This PR is a concrete protobuf interface for the stable gcs/autoscaler interface. This is just meant to be a concrete datapoint for discussion. There is no intention of merging this any time soon.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/protobuf/autoscaler.proto [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+  // Get the aggregated state of the cluster's nodes.
+  rpc GetAllNodeStatuses(GetAllNodeStatusesRequest) returns (GetAllNodeStatusesResponse);
+
+  // Set the node state for a batch of nodes. Nodes not included will retain their existing state.
+  rpc SetNodeState(SetNodeStateRequest) returns (SetNodeStateResponse);
+
+
+  // TODO: Do we need this? do we not? the world may never know.
+  rpc GetAllExplicitResourceRequests(GetAllExplicitResourceRequestsRequest) returns (GetAllExplicitResourceRequestsResponse);
```
Comment (wuisawesome) [inline src/ray/protobuf/autoscaler.proto:157]: In the original design doc this is handled in `GetAllQueuedResources` but it's really quite difficult/strange to have GCS to the binpacking for it...

File: src/ray/protobuf/autoscaler.proto [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+option cc_enable_arenas = true;
+package ray.rpc;
+
+message ResourceBundle {
+  // The fixed point resource size of the resource (granularity: 1/10000).
+  map<string, int64> resources = 1;
+}
+
+// A workload (task/actor/placement group) unfulfilled by the the cluster's internal scheduler(s).
+message QueuedResources {
```
Comment (wuisawesome) [inline src/ray/protobuf/autoscaler.proto:11]: Renamed `ResourceRequest` to this since "request" is too confusing/repetitive with grpc terminology.

File: src/ray/protobuf/autoscaler.proto [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+  // The number of resources of this quantity. This should be
+  // functionally equivalent to setting quantity=1 and repeating the
+  // field many times.
+  uint64 quantity = 1;
+
+  // How badly the cluster needs the resource 0-20. 20 means the
+  // cluster cannot complete the workload without satisfying the
+  // resource (e.g. a "hard" demand). 0 means the resource is only
+  // needed for convenience.
+  int32 necessity = 2;
```
Comment (DmitriGekhtman) [inline src/ray/protobuf/autoscaler.proto:38]: This might be one of the more controversial bits -- interpretation of necessity level is left to the scaling apparatus.
It's not 100% defined what's being conveyed. 

One way to go is to have an enum that conveys something about the nature of the scale request (e.g. "this request is for an actor")

I dunno, though.

File: src/ray/protobuf/autoscaler.proto (hunk 1 above)
Comment (DmitriGekhtman) [inline src/ray/protobuf/autoscaler.proto:157]: Would be nice if we didn't need it, but would take a bit of work to remove it 
(would need to move logic around sdk.request_resources resources from the autoscaler to the GCS)

Comment (DmitriGekhtman) [reply src/ray/protobuf/autoscaler.proto:157]: One way to do it without much refactor would be to express in the [enum](https://github.com/ray-project/ray/pull/23999#discussion_r852548503) suggested above that the request is an explicit sdk resource request.

Comment (DmitriGekhtman) [reply src/ray/protobuf/autoscaler.proto:157]: Hmm...still would need some GCS refactor to keep the permanence of the request...

File: src/ray/protobuf/autoscaler.proto [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+message ResourceBundle {
+  // The fixed point resource size of the resource (granularity: 1/10000).
+  map<string, int64> resources = 1;
+}
+
+// A workload (task/actor/placement group) unfulfilled by the the cluster's internal scheduler(s).
+message QueuedResources {
+
+  // Type of placement group strategy.
+  enum PlacementStrategy {
```
Comment (sriram-anyscale) [inline src/ray/protobuf/autoscaler.proto:14]: Could this enum be placed outside (i.e., as a peer of) QueuedResources?

File: src/ray/protobuf/autoscaler.proto [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+  uint64 quantity = 1;
+
+  // How badly the cluster needs the resource 0-20. 20 means the
+  // cluster cannot complete the workload without satisfying the
+  // resource (e.g. a "hard" demand). 0 means the resource is only
+  // needed for convenience.
+  int32 necessity = 2;
+
+  // Each individual resource bundle in the queued resources. For
+  // tasks or actors this is length 1.
```
Comment (sriram-anyscale) [inline src/ray/protobuf/autoscaler.proto:41]: Do you mean if this QueuedResource is for a task or an actor, then this field has one entry?
In what situations will it have more than one entry?

File: src/ray/protobuf/autoscaler.proto [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+  // TODO: Ideally we unify GCS's node id with the autoscaler's here.
+  // The ID of node.
+  bytes node_id = 2;
+
+  // The address of the node manager service for the node.
+  int32 node_manager_port = 3;
+  string node_manager_address = 4;
+
+  // The total fixed point resources (granularity 1/10000) of this node.
```
Comment (sriram-anyscale) [inline src/ray/protobuf/autoscaler.proto:83]: change "total" to "available"

File: src/ray/protobuf/autoscaler.proto [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+  // The state of the node.
+  NodeState node_state = 1;
+
+  // TODO: Ideally we unify GCS's node id with the autoscaler's here.
+  // The ID of node.
+  bytes node_id = 2;
+
+  // The address of the node manager service for the node.
+  int32 node_manager_port = 3;
+  string node_manager_address = 4;
```
Comment (sriram-anyscale) [inline src/ray/protobuf/autoscaler.proto:81]: Do we need the address and port?

File: src/ray/protobuf/autoscaler.proto [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+    // Places Bundles across distinct nodes.
+    // The group is not allowed to deploy more than one bundle on a node.
+    STRICT_SPREAD = 3;
+  }
+
+  // The number of resources of this quantity. This should be
+  // functionally equivalent to setting quantity=1 and repeating the
+  // field many times.
+  uint64 quantity = 1;
```
[the remark this answers is not in the document]
Comment (wuisawesome) [reply src/ray/protobuf/autoscaler.proto:32]: ambition ;) 

(i'll change it to 32 bit, but will save this comment when a user tries to queue up 4 billion tasks)

File: src/ray/protobuf/autoscaler.proto (hunk 3 above)
Comment (wuisawesome) [reply src/ray/protobuf/autoscaler.proto:38]: lemme try implementing the enum way.

File: src/ray/protobuf/autoscaler.proto (hunk 5 above)
Comment (wuisawesome) [reply src/ray/protobuf/autoscaler.proto:41]: it would have multiple entries for a placement group. let me rephrase the comment.

File: src/ray/protobuf/autoscaler.proto (hunk 1 above)
Comment (DmitriGekhtman) [reply src/ray/protobuf/autoscaler.proto:157]: We could keep this initially to keep code changes incremental.

File: src/ray/protobuf/autoscaler.proto (hunk 7 above)
Comment (wuisawesome) [reply src/ray/protobuf/autoscaler.proto:81]: We don't send requests using this information, but we do display it for observability information.

Comment (wuisawesome) [reply src/ray/protobuf/autoscaler.proto:81]: removing port since i guess we only display address (and port could be encoded in that if we need to)

File: src/ray/protobuf/autoscaler.proto (hunk 1 above)
Comment (wuisawesome) [reply src/ray/protobuf/autoscaler.proto:157]: ohhhh that's clever. part of it feels a little weird, but i like it more than introducing a new concept. let me do that

File: src/ray/protobuf/autoscaler.proto (hunk 6 above)
Comment (wuisawesome) [reply src/ray/protobuf/autoscaler.proto:118]: assuming this comment was intended for the comment over `resources_available` which I updated.

File: src/ray/protobuf/autoscaler.proto [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+  FEASIBLE_TASK = 0;
+
+  // A task that the cluster does not have the correct resources to fulfill.
+  INFEASIBLE_TASK = 1;
+
+  // A actor that (assuming no deadlock can eventually be scheduled on the cluster.
+  FEASIBLE_ACTOR = 2;
+
+  // A actor that the cluster does not have the correct resources to fulfill.
+  INFEASIBLE_ACTOR = 3;
```
Comment (ericl) [inline src/ray/protobuf/autoscaler.proto:46]: Why not use a bool for feasible instead of adding the full enum cross product?

File: src/ray/protobuf/autoscaler.proto [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+// A workload (task/actor/placement group) unfulfilled by the the cluster's internal scheduler(s).
+message QueuedResources {
+  // The number of resources of this quantity. This should be
+  // functionally equivalent to setting quantity=1 and repeating the
+  // field many times.
+  uint32 quantity = 1;
+
+  // The queue type this resource comes from.
+  QueueType type = 2;
```
Comment (ericl) [inline src/ray/protobuf/autoscaler.proto:69]: Queue is an implementation detail-- can call this source/info.

**Comment by scv119** [comment]
Does it make sense to link the design doc?

File: src/ray/protobuf/autoscaler.proto [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+// A workload (task/actor/placement group) unfulfilled by the the cluster's internal scheduler(s).
+message QueuedResources {
+  // The number of resources of this quantity. This should be
+  // functionally equivalent to setting quantity=1 and repeating the
+  // field many times.
+  uint32 quantity = 1;
+
+  // Whether the queued resources can eventually be scheduled on the cluster
+  // (neglecting deadlock).
+  bool is_feasible = 2;
```
Comment (ericl) [inline src/ray/protobuf/autoscaler.proto:61]:
```suggestion
  bool is_currently_feasible = 2;
```

To clarify, this is based on current node types right? The request could eventually become feasible if the autoscaler adds future nodes to the cluster.

Comment (ericl) [reply src/ray/protobuf/autoscaler.proto:61]: It would also be good to add what the use for this is--- at first glance it's a bit odd to include in the interface.

**Comment by wuisawesome** [comment]
> Does it make sense to link the design doc?

I don't think we have a public design doc. Perhaps I should use this as an opportunity to create an REP...

File: src/ray/protobuf/autoscaler.proto [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+
+  // Represents a job's call to
+  // `ray.autoscaler.sdk.request_resources()`. This type requests
+  // resources to be available on the cluster but won't actually
+  // utilize them.
+  SDK_RESOURCE_REQUEST = 3;
+}
+
+// A workload (task/actor/placement group) unfulfilled by the the cluster's internal scheduler(s).
+message QueuedResources {
```
Comment (scv119) [inline src/ray/protobuf/autoscaler.proto:53]: QueuedResourceDemands? or just ResourceDemands?

File: src/ray/protobuf/autoscaler.proto [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+option cc_enable_arenas = true;
+package ray.rpc;
+
+message ResourceBundle {
+  // The fixed point resource size of the resource (granularity: 1/10000).
+  map<string, int64> resources = 1;
+}
+
+// Type of placement group strategy.
+enum PlacementStrategy {
```
Comment (scv119) [inline src/ray/protobuf/autoscaler.proto:20]: task/actor has PlacementStrategy too. i.e. NodeAffinityStrategy. What's our thought here?

File: src/ray/protobuf/autoscaler.proto [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+*/
+
+
+syntax = "proto3";
+option cc_enable_arenas = true;
+package ray.rpc;
+
+message ResourceBundle {
+  // The fixed point resource size of the resource (granularity: 1/10000).
+  map<string, int64> resources = 1;
```
Comment (scv119) [inline src/ray/protobuf/autoscaler.proto:16]: does unit resource instances apply here?

File: src/ray/protobuf/autoscaler.proto [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+  // The source of this queued resource.
+  SourceType type = 3;
+
+  // If the queued resource represents an actor or task, there is
+  // 1 entry in `bundles`. In a placement group, each bundle
+  // receives its own entry.
+  repeated ResourceBundle bundles = 4;
+
+  // The placement strategy of bundles within this queued resources.
+  PlacementStrategy placement_strategy = 5;
```
Comment (scv119) [inline src/ray/protobuf/autoscaler.proto:72]: do we care about pending time?

File: src/ray/protobuf/autoscaler.proto (hunk 12 above)
Comment (wuisawesome) [reply src/ray/protobuf/autoscaler.proto:53]: I'm fine with that as long as you're ok with the name _almost_ conflicting with the internal `ResourceDemand` message.

File: src/ray/protobuf/autoscaler.proto (hunk 13 above)
Comment (wuisawesome) [reply src/ray/protobuf/autoscaler.proto:20]: In the specific case of `NodeAffinityStrategy` (and the other existing strategies), I don't think the autoscaler can really do much with that information, but in general, I think that's a very good question and I don't have a general answer...

Do you have any specific scheduling strategies in mind?

File: src/ray/protobuf/autoscaler.proto (hunk 14 above)
Comment (wuisawesome) [reply src/ray/protobuf/autoscaler.proto:16]: No, this is a bug in the current autoscaler too.  Actually, I think this field is fine, (you can't do `num_gpus=[0.2, 0.3]`), but for the node resource usage, do you think we should make them lists of available? (it _feels_ kinda bad, but I don't see a better option if we want to fix that bug).

File: src/ray/protobuf/autoscaler.proto (hunk 15 above)
Comment (wuisawesome) [reply src/ray/protobuf/autoscaler.proto:72]: the autoscaler currently doesn't have/use that information, but if you think it's good for future proofing, we can add it.

File: src/ray/protobuf/autoscaler.proto (hunk 13 above)
Comment (DmitriGekhtman) [reply src/ray/protobuf/autoscaler.proto:20]: Maybe the comment should say "placement strategy" (drop the "group")?

I have not been following the progress on the new placement APIs at all, but is there overlap with the placement groups concepts?
If not, is there potential for useful concept/code overlap?

File: src/ray/protobuf/autoscaler.proto [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+
+  // The source of this queued resource.
+  SourceType type = 3;
+
+  // If the queued resource represents an actor or task, there is
+  // 1 entry in `bundles`. In a placement group, each bundle
+  // receives its own entry.
+  repeated ResourceBundle bundles = 4;
+
+  // The placement strategy of bundles within this queued resources.
```
Comment (DmitriGekhtman) [inline src/ray/protobuf/autoscaler.proto:71]: Grammar: "this queued resources"

File: src/ray/protobuf/autoscaler.proto (hunk 13 above)
Comment (wuisawesome) [reply src/ray/protobuf/autoscaler.proto:20]: So the SchedulingStrategies are for tasks/actors (with perhaps an overlapping "job to be done" with placement groups). At the moment,

* Spread is soft so we can ignore it while binpakcing
* Node affinity can be filtered within GCS
* Default is default
* Placement group - these require the special case handling.

The most likely thing that could really throw a wrench in this is some soft of Node-anti-affinity strategy, or placement group resizing, or placement group rescheduling.

Comment (scv119) [reply src/ray/protobuf/autoscaler.proto:20]: cc @jjyao @rkooo567

File: src/ray/protobuf/autoscaler.proto (hunk 14 above)
Comment (DmitriGekhtman) [reply src/ray/protobuf/autoscaler.proto:16]: Can you say more about what the bug is?

File: src/ray/protobuf/autoscaler.proto [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+  // Places Bundles across distinct nodes.
+  // The group is not allowed to deploy more than one bundle on a node.
+  STRICT_SPREAD = 3;
+}
+
+enum SourceType {
+  // A task that (assuming no deadlock can eventually be scheduled on the cluster.
+  TASK = 0;
+
+  // A actor that (assuming no deadlock can eventually be scheduled on the cluster.
```
Comment (DmitriGekhtman) [inline src/ray/protobuf/autoscaler.proto:39]: Comment missing a paren.

There are some proponents (including me) of considering actor resources demands "hard" (should trigger immediate upscaling).

The advantage of including the source type is that the autoscaler implementation can make that call.

File: src/ray/protobuf/autoscaler.proto [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+  ACTOR = 1;
+
+  // A placement group that (assuming no deadlock can eventually be scheduled on the cluster.
+  PLACEMENT_GROUP = 2;
+
+  // Represents a job's call to
+  // `ray.autoscaler.sdk.request_resources()`. This type requests
+  // resources to be available on the cluster but won't actually
+  // utilize them.
```
Comment (DmitriGekhtman) [inline src/ray/protobuf/autoscaler.proto:48]: Noting again we'd have to make sure this kind of request is persisted until overwritten by another request_resources call.

File: src/ray/protobuf/autoscaler.proto [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+
+// A workload (task/actor/placement group) unfulfilled by the the cluster's internal scheduler(s).
+message QueuedResources {
+  // The number of resources of this quantity. This should be
+  // functionally equivalent to setting quantity=1 and repeating the
+  // field many times.
+  uint32 quantity = 1;
+
+  // Whether the queued resources can eventually be scheduled on the cluster
+  // (neglecting deadlock).
```
Comment (DmitriGekhtman) [inline src/ray/protobuf/autoscaler.proto:60]: Is this right --
neglecting deadlock == could be scheduled somewhere assuming the current collection of nodes and 0% utilization of those nodes

File: src/ray/protobuf/autoscaler.proto (hunk 11 above)
Comment (DmitriGekhtman) [reply src/ray/protobuf/autoscaler.proto:61]: This one is tricky. It's like the scheduler saying "I can get to this eventually so you don't really need to scale."
On the other hand "neglecting deadlock" indicates that the scheduler can't have this knowledge.

I'd lean towards leaving the feasibility judgement to the autoscaler for now.
(That's how we currently have it.)

File: src/ray/protobuf/autoscaler.proto (hunk 15 above)
Comment (DmitriGekhtman) [reply src/ray/protobuf/autoscaler.proto:72]: Ooh, yeah, how long a demand has been pending would be a great thing to take into account when autoscaling.
IMO, better left for later.

File: src/ray/protobuf/autoscaler.proto [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+    DRAINING = 2;
+
+    // The node is no longer active in the cluster it is safe to remove it.
+    DEAD = 3;
+  }
+
+  // The state of the node.
+  NodeState node_state = 1;
+
+  // TODO: Ideally we unify GCS's node id with the autoscaler's here.
```
Comment (DmitriGekhtman) [inline src/ray/protobuf/autoscaler.proto:102]: Cloud provider identity of a node is always going to be a different thing.
But we could probably improve how the two identities are managed.

File: src/ray/protobuf/autoscaler.proto [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+
+  // The underlying host type (e.g. instance type or machine type).
+  string host_type = 6;
+}
+
+message GetAllNodeStatusesRequest {
+}
+
+message GetAllNodeStatusesResponse {
+  // The set of all nodes that ever were or ever will be in the cluster.
```
Comment (DmitriGekhtman) [inline src/ray/protobuf/autoscaler.proto:124]: well..."will be" that we know of

Comment is clear already,
but something like "Set of all past, present, and pending nodes" is more precise.

File: src/ray/protobuf/autoscaler.proto [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+  ERROR = 2;
+}
+
+message PublishAutoscalerEventRequest {
+
+  // The event level of the autoscaler event.
+  AutoscalerEventLevel level = 1;
+
+  // Autoscaler event type.
+  string event_type = 2;
```
Comment (DmitriGekhtman) [inline src/ray/protobuf/autoscaler.proto:151]: Is an enum better style for this one?

File: src/ray/protobuf/autoscaler.proto [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+
+  // The autoscaler event message.
+  string message = 3;
+}
+
+message PublishAutoscalerEventResponse {
+  // True if GCS now claims respnsibility for communicating this event.
```
Comment (DmitriGekhtman) [inline src/ray/protobuf/autoscaler.proto:158]: Typo: respnsibility
Could you clarify the comment?

**Review by DmitriGekhtman — DISMISSED** [review]
Approving so that we don't get the impression that we're paralyzed on this.

File: src/ray/protobuf/autoscaler.proto (hunk 13 above)
Comment (jjyao) [reply src/ray/protobuf/autoscaler.proto:20]: I think we should send SchedulingStrategies information to autoscaler as well so it has enough information to make a good decision. NodeAntiAffinityStrategy would be interesting and the whole taint and tolerance thing.

File: src/ray/protobuf/autoscaler.proto (hunk 22 above)
Comment (wuisawesome) [reply src/ray/protobuf/autoscaler.proto:151]: I would actually advocate for an opaque string field here since the autoscaler is sending this message. it's more difficult for a newer autoscaler to send a newer protocol to gcs than to receive a message with an older protocol.

Comment (DmitriGekhtman) [reply src/ray/protobuf/autoscaler.proto:151]: Ok -- if we're just using this to format status messages, then it makes sense to keep this a string and use a Python enum in the autoscaler code.

File: src/ray/protobuf/autoscaler.proto (hunk 11 above)
Comment (DmitriGekhtman) [reply src/ray/protobuf/autoscaler.proto:61]: So, given a single bundle, feasible would mean that 
the node type of some currently active node has capacity for the bundle.

Given a placement group of resources, it would mean something more complex.

I think feasible is slippery to define but given a definition straightforward to compute in either the GCS or autoscaler.

File: src/ray/protobuf/autoscaler.proto [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+  repeated QueuedResources queued_resources = 1;
+}
+
+message NodeInfo {
+  // State of a node.
+  enum NodeState {
+    // The node has not joined the cluster yet.
+    PENDING = 0;
+
+    // The node has joined the cluser and is healthy.
```
Comment (DmitriGekhtman) [inline src/ray/protobuf/autoscaler.proto:89]: typo: cluster

File: src/ray/protobuf/autoscaler.proto (hunk 20 above)
Comment (DmitriGekhtman) [reply src/ray/protobuf/autoscaler.proto:102]: > Cloud provider identity of a node is always going to be a different thing.

Or maybe not -- SetNodeInfo seems a fundamental change to the interface.
Looks like the autoscaler is now telling the GCS what the name of the node is, so we can pass the cloud provider name of the node straight to the the GCS?

Comment (DmitriGekhtman) [reply src/ray/protobuf/autoscaler.proto:102]: err...or maybe there's a typo below

File: src/ray/protobuf/autoscaler.proto [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+message GetAllNodeStatusesResponse {
+  // The set of all nodes that ever were or ever will be in the cluster.
+  repeated NodeInfo all_nodes = 1;
+}
+
+message SetNodeStateRequest {
+  // Set a batch of node states.
+  repeated NodeInfo batch_to_update = 1;
```
Comment (DmitriGekhtman) [inline src/ray/protobuf/autoscaler.proto:130]: Should this be `NodeState` rather than `NodeInfo`?

File: src/ray/protobuf/autoscaler.proto (hunk 20 above)
Comment (DmitriGekhtman) [reply src/ray/protobuf/autoscaler.proto:102]: Disregard above and see question about NodeState and NodeInfo below.

File: src/ray/protobuf/autoscaler.proto (hunk 25 above)
Comment (DmitriGekhtman) [reply src/ray/protobuf/autoscaler.proto:130]: I think I understand what it means to set the NodeState to `Draining` (it's a command to drain the node?)
What does it mean for the autoscaler to set the NodeState to `Alive`?

File: src/ray/protobuf/autoscaler.proto (hunk 20 above)
Comment (DmitriGekhtman) [reply src/ray/protobuf/autoscaler.proto:102]: Eh, or don't disregard.
It looks like we have a bit of a hanging question from the original about how exactly `SetNodeState` is supposed to work.
https://docs.google.com/document/d/197P0Ia_dPbHgfvJbo0gauP1BhPrlEheMibrnZjLhOUU/edit?disco=AAAAVYH-xe0

**Comment by DmitriGekhtman** [comment]
unstale
